### PR TITLE
[WIP] High precision simulator comparison testing of section 5 calculations

### PR DESCRIPTION
**Pull request by relyt29**
## What is the purpose of the change

Adds unit tests to confirm that implementation of section 5 matches research simulations.

## Testing and Verifying

This change introduces new tests in emissions_test.go

## Documentation and Release Note

no Documentation changes.

### CODE REVIEW

File: test/testutil/inEpsilon.go [hunk 1]
```diff
@@ -45,5 +63,11 @@ func InEpsilon4(t *testing.T, value alloraMath.Dec, target string) {
 }*/
 
 func InEpsilon5(t *testing.T, value alloraMath.Dec, target string) {
-	InEpsilon(t, value, target, "0.00001")
+	epsilonDec := alloraMath.MustNewDecFromString("0.00001")
+	targetDec := alloraMath.MustNewDecFromString(target)
+	InEpsilon(t, value, targetDec, epsilonDec)
+}
+
+func InEpsilon5D(t *testing.T, value alloraMath.Dec, target alloraMath.Dec) {
```
Comment (guilherme-brandao) [inline test/testutil/inEpsilon.go:71]: `inEpsilon5Dec` ?

File: x/emissions/keeper/keeper.go [hunk 2]
```diff
@@ -1058,7 +1058,7 @@ func (k *Keeper) RemoveReputerStake(
 //	stakeSumFromDelegator, delegatedStakes, stakeFromDelegatorsUponReputer
 func (k *Keeper) RemoveDelegateStake(
 	ctx context.Context,
-	blockHeight BlockHeight,
+	stakeRemovalHeight BlockHeight,
```
Comment (guilherme-brandao) [inline x/emissions/keeper/keeper.go:1061]: `stakeRemovalBlockHeight` ?

File: x/emissions/keeper/keeper.go [hunk 3]
```diff
@@ -1498,26 +1512,40 @@ func (k Keeper) GetDelegateStakeRemoval(
...
 	ret := make([]types.DelegateStakeRemovalInfo, 0)
-	rng := NewSinglePrefixedQuadrupleRange[BlockHeight, TopicId, ActorId, ActorId](blockHeight)
+
+	// make a range that has everything less than the block height, inclusive
+	startKey := QuadrupleSinglePrefix[BlockHeight, TopicId, ActorId, ActorId](0)
+	rng := &collections.Range[Quadruple[BlockHeight, TopicId, ActorId, ActorId]]{}
+	rng = rng.Prefix(startKey)
+	endKey := QuadrupleSinglePrefix[BlockHeight, TopicId, ActorId, ActorId](blockHeight + 1)
+	rng = rng.EndExclusive(endKey)
+
```
Comment (guilherme-brandao) [inline x/emissions/keeper/keeper.go:1530]: same here(?)

File: x/emissions/module/abci.go [hunk 4]
```diff
@@ -15,10 +15,14 @@ func EndBlocker(ctx context.Context, am AppModule) error {
...
+	moduleParams, err := am.keeper.GetParams(sdkCtx)
+	if err != nil {
+		sdkCtx.Logger().Error("Error Getting module params", err)
+		return err
+	}
 	// Remove Stakers that have been wanting to unstake this block. They no longer get paid rewards
-	RemoveStakes(sdkCtx, blockHeight, am.keeper)
-	RemoveDelegateStakes(sdkCtx, blockHeight, am.keeper)
+	RemoveStakes(sdkCtx, blockHeight, am.keeper, moduleParams.HalfMaxProcessStakeRemovalsEndBlock)
+	RemoveDelegateStakes(sdkCtx, blockHeight, am.keeper, moduleParams.HalfMaxProcessStakeRemovalsEndBlock)
```
Comment (guilherme-brandao) [inline x/emissions/module/abci.go:25]: Add error handling

**Review by guilherme-brandao — COMMENTED** [review]
Just minor things/ suggestions for now! Looking good

File: x/emissions/module/abci.go (hunk 4 above)
Comment (relyt29) [reply x/emissions/module/abci.go]: intentionally does not throw errors, so as not to affect consensus

also not relevant for this PR, should have brought this up in #469

File: x/emissions/keeper/keeper.go [hunk 5]
```diff
@@ -1393,26 +1393,40 @@ func (k Keeper) GetStakeRemoval(
...
+	limit uint64,
+) (ret []types.StakeRemovalInfo, anyLeft bool, err error) {
+	ret = make([]types.StakeRemovalInfo, 0)
+	// make a range that has everything less than the block height, inclusive
+	startKey := collections.TriplePrefix[BlockHeight, TopicId, ActorId](0)
+	rng := &collections.Range[collections.Triple[BlockHeight, TopicId, ActorId]]{}
+	rng = rng.Prefix(startKey)
+	// +1 for end exclusive. Don't know why end inclusive is being buggy but it is
+	endKey := collections.TriplePrefix[BlockHeight, TopicId, ActorId](blockHeight + 1)
+	rng = rng.EndExclusive(endKey)
```
[the remark this answers is not in the document]
Comment (relyt29) [reply x/emissions/keeper/keeper.go]: I'm afraid lol

**Review by kpeluso — APPROVED** [review]

**Pull request merged**